File: Platforms/Common/Reign.Video.D3D11/Shaders/ShaderResource.cs
using System;
using Reign.Core;
using Reign.Video;
using Reign_Video_D3D11_Component;

namespace Reign.Video.D3D11
{
	public class ShaderResource : IShaderResource
	{
		#region Properties
		private ShaderResourceCom com;
		public string Name {get; private set;}
		#endregion

		#region Constructors
		public ShaderResource(string name, VideoCom video, ShaderModelCom vertexShader, ShaderModelCom pixelShader, int vertexIndex, int pixelIndex)
		{
			Name = name;
			com = new ShaderResourceCom(video, vertexShader, pixelShader, vertexIndex, pixelIndex);
		}
		#endregion

		#region Methods
		public void Set(ITexture2D resource)
		{
			com.Set(((Texture2D)resource).com);
		}

		public void Set(ITexture3D resource)
		{
			throw new NotImplementedException();
		}
		#endregion
	}
}

File: Platforms/Common/Reign_Video_D3D11_Component/Shaders/ShaderResource.cpp
#include "pch.h"
#include "ShaderResource.h"

namespace Reign_Video_D3D11_Component
{
	#pragma region Constructors
	ShaderResourceCom::ShaderResourceCom(VideoCom^ video, ShaderModelCom^ vertexShader, ShaderModelCom^ pixelShader, int vertexIndex, int pixelIndex)
	{
		this->video = video;
		this->vertexResources = vertexShader->resources;
		this->pixelResources = pixelShader->resources;
		this->vertexIndex = vertexIndex;
		this->pixelIndex = pixelIndex;
	}
	#pragma endregion

	#pragma region Methods
	void ShaderResourceCom::Set(Texture2DCom^ resource)
	{
		if (vertexIndex != -1)
		{
			vertexResources[vertexIndex] = resource->shaderResource;
			video->currentVertexResources[vertexIndex] = resource->shaderResource;
		}
		if (pixelIndex != -1)
		{
			pixelResources[pixelIndex] = resource->shaderResource;
			video->currentPixelResources[pixelIndex] = resource->shaderResource;
		}
	}

	/*void ShaderResource::Set(Texture3DI^ resource)
	{
		Texture3D^ texture = (Texture3D^)resource;
		if (vertexIndex != -1)
		{
			vertexResources[vertexIndex] = texture->ShaderResource;
			video->currentVertexResources[vertexIndex] = texture->ShaderResource;
		}
		if (pixelIndex != -1)
		{
			pixelResources[pixelIndex] = texture->ShaderResource;
			video->currentPixelResources[pixelIndex] = texture->ShaderResource;
		}
	}*/
	#pragma endregion
}